### PR TITLE
add support for gdb path substitution

### PR DESCRIPTION
**Pull request by karljs**
When debugging a binary, if the build environment is different from the debugging environment (e.g., built in Docker, debugging locally) you might see errors where the debugger isn't able to find the source files that match up with the binary.  This manifests in the extension as error messages in the debug console along the lines of "No source file named ...".

GDB solves this problem with "path substitutions" in a variable.  So if you build a binary in `/root` and want to debug it in `/home/username` you need to issue the command `set substitute-path /root /home/username`. 

This PR adds a configuration option for `launch.json` that lets you set this.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
 		"native",
 		"debug"
 	],
-	"version": "0.24.0",
+	"version": "0.25.0",
```
Comment (WebFreak001) [inline package.json:12]: please don't change the version property (I change it on release when I call vsce and make it a separate commit)

File: src/backend/mi2/mi2.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ function couldBeOutput(line: string) {
 const trace = false;
 
 export class MI2 extends EventEmitter implements IBackend {
-	constructor(public application: string, public preargs: string[], public extraargs: string[], procEnv: any) {
+	constructor(public application: string, public preargs: string[], public extraargs: string[], procEnv: any, public pathSubstitutions: any = {}) {
```
Comment (WebFreak001) [inline src/backend/mi2/mi2.ts:30]: to avoid bloating this method any more in future changes, can you make this a `public extraCommands: string[] = []`?

File: src/backend/mi2/mi2.ts [hunk 3]
```diff
@@ -197,6 +197,10 @@ export class MI2 extends EventEmitter implements IBackend {
 			cmds.push(this.sendCommand("file-exec-and-symbols \"" + escape(target) + "\""));
 		if (this.prettyPrint)
 			cmds.push(this.sendCommand("enable-pretty-printing"));
+		
+		for (const key in this.pathSubstitutions) {
+			cmds.push(this.sendCommand("gdb-set substitute-path " + key + " " + this.pathSubstitutions[key]));
+		};
```
Comment (WebFreak001) [inline src/backend/mi2/mi2.ts:203]: this logic would then be moved into GDB, or if this also works with LLDB put this into a function in MI2 which then modifies `this.extraCommands` and is called in gdb.ts before initDebugger

Instead it would here just do
```ts
this.extraCommands.forEach((cmd) => {
	cmds.push(this.sendCommand(cmd));
});
```

File: package.json [hunk 4]
```diff
@@ -107,6 +107,13 @@
 								"description": "Additional arguments to pass to GDB",
 								"default": []
 							},
+							"pathSubstitutions": {
+								"type": "object",
+								"description": "Help GDB find your source using path substitutions (GDB `substitute-path` variable",
+								"default": {
+									"<fromPath>": "<toPath>"
+								}
+							},
```
Comment (WebFreak001) [inline package.json:116]: you need to copy this to the attach arguments too

File: src/gdb.ts [hunk 5]
```diff
@@ -10,6 +10,7 @@ export interface LaunchRequestArguments extends DebugProtocol.LaunchRequestArgum
 	gdbpath: string;
 	env: any;
 	debugger_args: string[];
+	pathSubstitutions: any;
```
Comment (WebFreak001) [inline src/gdb.ts:13]: make this a `{[index: string]: string}` type

File: package.json (hunk 1 above)
Comment (karljs) [reply package.json]: Reverted that, sorry.

**Comment by karljs** [comment]
Thanks for the suggestions.  I've made them all, and also added support for LLDB (with the caveat that I don't currently have it set up and can't test it thoroughly.)

File: package.json [hunk 6]
```diff
@@ -107,6 +107,13 @@
 								"description": "Additional arguments to pass to GDB",
 								"default": []
 							},
+							"pathSubstitutions": {
+								"type": "object",
+								"description": "Help GDB find your source using path substitutions (GDB `substitute-path` variable",
```
Comment (WebFreak001) [inline package.json:112]: you didn't close the parentheses

File: package.json [hunk 7]
```diff
@@ -252,6 +259,13 @@
...
```
Comment (WebFreak001) [inline package.json:264]: ditto

File: package.json [hunk 8]
```diff
@@ -515,6 +529,13 @@
 								"description": "Additional arguments to pass to LLDB",
 								"default": []
 							},
+							"pathSubstitutions": {
+								"type": "object",
+								"description": "Help LLDB find your source using path substitutions (LLDB `target.source-map` variable",
```
Comment (WebFreak001) [inline package.json:534]: ditto

File: package.json [hunk 9]
```diff
@@ -655,6 +676,13 @@
...
```
Comment (WebFreak001) [inline package.json:681]: ditto

**Comment by karljs** [comment]
Good catch, thanks.  Fixed the typos
